### PR TITLE
Cover gaps in reduction tests

### PR DESCRIPTION
**Pull request by aleksmesh**
Add test cases for skipped reducer combines.

### CODE REVIEW

**Comment by aleksmesh** [comment]
@bader CI failed  during build redaction group tests due to large amount of RAM required for the tests to build.
We will try to split the source files into a larger number of test cases.

**Comment by TApplencourt** [comment]
Talking about gat, I think We need to test reduction on `host`, `device`, and `shared`  USM allocator too.  I think we testing only `shared`.

I think this may discover some gap in the spec (aka, if reduction use atomic maybe we can do it on the host memory from the device if they don't support this kind of atomic)

**Review by keryell — COMMENTED** [review]
This is massive! :-)

File: tests/reduction/reducer_api.h [hunk 1]
```diff
@@ -69,6 +69,14 @@ struct check_reducer_subscript {
                 acc_results[i++] =
                     std::is_same_v<OperatorT, typename red_t::binary_operation>;
                 acc_results[i++] = 0 == red_t::dimensions;
+                AccumulatorT val{1};
+                auto& res_combine = reducer.combine(val);
+                acc_results[i++] =
+                    std::is_same_v<decltype(res_combine), decltype(reducer)>;
+                // OperatorT is sycl::plus, so checking the result of += method
```
[the remark this answers is not in the document]
Comment (vladimirkhashev) [reply tests/reduction/reducer_api.h]: Replaced.

File: tests/reduction/reducer_api.h [hunk 2]
```diff
@@ -105,6 +113,14 @@ struct check_reducer_subscript {
...
```
[the remark this answers is not in the document]
Comment (vladimirkhashev) [reply tests/reduction/reducer_api.h]: Replaced.

File: tests/reduction/reduction_common.h [hunk 3]
```diff
@@ -47,13 +68,37 @@ const auto scalar_types =
...
   sycl::host_accessor buf_accessor{buffer};
-  return std::accumulate(buf_accessor.begin(), buf_accessor.end(),
-                         value_for_initialization, functor);
+  if constexpr (TestCaseT == test_case_type::each_work_item) {
+    return std::accumulate(buf_accessor.begin(), buf_accessor.end(),
+                           value_for_initialization, functor);
+  } else if constexpr (TestCaseT == test_case_type::each_even_work_item) {
+    int counter = 1;
+    auto it = buf_accessor.begin();
+    while (it != buf_accessor.end()) {
```
[the remark this answers is not in the document]
Comment (vladimirkhashev) [reply tests/reduction/reduction_common.h]: Changed to ranged `for()`.

File: tests/reduction/reduction_common.h [hunk 4]
```diff
@@ -47,13 +68,37 @@ const auto scalar_types =
...
+      }
+      ++it;
+      ++counter;
+    }
+    return expected_value;
+  } else if constexpr (TestCaseT == test_case_type::no_one_work_item) {
+    return expected_value;
+  } else if constexpr (TestCaseT == test_case_type::each_work_item_twice) {
+    auto it = buf_accessor.begin();
+    while (it != buf_accessor.end()) {
```
[the remark this answers is not in the document]
Comment (vladimirkhashev) [reply tests/reduction/reduction_common.h]: Changed to ranged `for()`.

File: tests/reduction/reduction_get_lambda.h [hunk 5]
```diff
@@ -87,6 +101,83 @@ auto get_lambda_with_range(AccessorT accessor) {
...
+ *  @tparam AccessorT buffer accessor type
+ *  @param accessor Accessor to the buffer
+ *  @param number_elements Number elements in sycl::span, used only when
+ *         constructing reducer with span
+ *  @retval Lambda with chosen operator
+ */
+template <typename VariableT, typename FunctorT, bool UseCombineFlagT,
+          typename AccessorT>
+auto get_lambda_with_range_even_item(AccessorT accessor) {
+  if constexpr (UseCombineFlagT) {
```
[the remark this answers is not in the document]
Comment (vladimirkhashev) [reply tests/reduction/reduction_get_lambda.h]: Moved.

File: tests/reduction/reduction_get_lambda.h [hunk 6]
```diff
@@ -87,6 +101,83 @@ auto get_lambda_with_range(AccessorT accessor) {
...
+ *  @param number_elements Number elements in sycl::span, used only when
+ *         constructing reducer with span
+ *  @retval Lambda with chosen operator
+ */
+template <typename VariableT, typename FunctorT, bool UseCombineFlagT,
+          typename AccessorT>
+auto get_lambda_with_range_item_twice(AccessorT accessor) {
+  if constexpr (UseCombineFlagT) {
+    return [=](sycl::id<1> idx, auto &reducer) {
+      reducer.combine(accessor[idx]).combine(accessor[idx]);
```
[the remark this answers is not in the document]
Comment (vladimirkhashev) [reply tests/reduction/reduction_get_lambda.h:163]: Moved.

File: tests/reduction/reduction_get_lambda.h [hunk 7]
```diff
@@ -114,6 +205,90 @@ auto get_lambda_with_nd_range(AccessorT accessor, size_t number_elements = 0) {
...
+ */
+template <typename VariableT, typename FunctorT, bool UseCombineFlagT,
+          typename AccessorT>
+auto get_lambda_with_nd_range_even_item(AccessorT accessor,
+                                        size_t number_elements = 0) {
+  if constexpr (UseCombineFlagT) {
+    return [=](sycl::nd_item<1> nd_item, auto &reducer) {
+      size_t num = nd_item.get_global_id();
+      if (num & 1) {
+        reducer.combine(accessor[nd_item.get_global_id()]);
```
[the remark this answers is not in the document]
Comment (vladimirkhashev) [reply tests/reduction/reduction_get_lambda.h:215]: Done.

**Review by keryell — APPROVED** [review]
Thanks for the clarification.

**Pull request merged**